### PR TITLE
Add magnetometer thermal calibration/compensation.

### PR DESCRIPTION
**Pull request by mcsauder**
## Describe problem solved by this pull request
This PR extends the thermal compensation to include magnetometers.  Sensor ordering following the convention `accel, gyro, mag, baro` is also standardized across files in this PR.

## Describe your solution
Data taken from an IST8310 magnetometer standard with the pixhawk4 and pixhawk4 mini shows a change in measured values from 0C to 45C by almost ~10uTesla (0.1Gauss).  The typical earth magnetic field strength is ~60uTesla, so the observed shift is significant against what is being measured and more significant in areas of the earth with weaker magnetic field strength.
image

## Test data / coverage
Flight tested with calibration values applied at ambient (~21C): https://review.px4.io/plot_app?log=6335c430-828b-4f7e-8404-ba7832585ada
After cold soaking to ~7C and temperature climbing to ~17C: https://review.px4.io/plot_app?log=6335c430-828b-4f7e-8404-ba7832585ada 

## Additional context
Most magnetometers do not have an integrated temperature sensor, so this PR utilizes the primary barometer temperature value as the magnetometer temperature.  If there is a better alternative we can certainly use a different source (such as the IMU).

Let me know if you have questions or feedback on this PR. Thanks!

-Mark

### CODE REVIEW

**Comment by dagar** [comment]
Overall this looks good, but I really don't think we should be plumbing in random baro temperature in a magnetometer driver just for a temperature.

Should we do something to make the temperature source configurable (at the level of the compensation)?

**Comment by mcsauder** [comment]
>Should we do something to make the temperature source configurable (at the level of the compensation)?

Yes.  Good suggestion.  I'll work on that.  Thanks @dagar !

File: src/drivers/magnetometer/isentek/ist8308/IST8308.hpp [hunk 1]
```diff
@@ -91,17 +102,14 @@ class IST8308 : public device::I2C, public I2CSPIDriver<IST8308>
 
 	hrt_abstime _reset_timestamp{0};
 	hrt_abstime _last_config_check_timestamp{0};
-	int _failure_count{0};
+	hrt_abstime _temperature_update_timestamp{0};
```
Comment (dagar) [inline src/drivers/magnetometer/isentek/ist8308/IST8308.hpp:105]: Can this be reverted now?

**Comment by dagar** [comment]
Overall looks fine, but I do have a few suggestions for later

 - Tools/process_sensor_caldata.py
     - remove initial (and probably final) garbage data
     - decode device ids
     - verify onboard cal gives same results as offline script
  - split sensor_corrections for accel, gyro, baro, mag and publish per instance on temperature change
  - calibration only when the vehicle/board is still (conservative monitoring of accel/gyro)
  - heavily filter onboard calibration raw sensor data

File: src/drivers/magnetometer/isentek/ist8310/IST8310.hpp [hunk 2]
```diff
@@ -46,6 +46,7 @@
 #include <lib/drivers/device/i2c.h>
 #include <lib/drivers/magnetometer/PX4Magnetometer.hpp>
 #include <lib/perf/perf_counter.h>
+#include <lib/sensor_calibration/Barometer.hpp>
```
Comment (dagar) [inline src/drivers/magnetometer/isentek/ist8310/IST8310.hpp:49]:
```suggestion
```

File: src/lib/sensor_calibration/Magnetometer.hpp [hunk 3]
```diff
@@ -98,15 +97,21 @@ class Magnetometer
...
 
 private:
+	uORB::Subscription _sensor_correction_sub{ORB_ID(sensor_correction)};
+
 	Rotation _rotation_enum{ROTATION_NONE};
 
 	matrix::Dcmf _rotation;
 	matrix::Vector3f _offset;
 	matrix::Matrix3f _scale;
+	matrix::Vector3f _thermal_offset;
```
Comment (dagar) [inline src/lib/sensor_calibration/Magnetometer.hpp:112]: Make sure this field is reset with the rest. Maybe try side by side comparing Magnetometer.cpp and Accelerometer.cpp to keep them mostly in sync.

File: src/lib/sensor_calibration/Magnetometer.hpp [hunk 4]
```diff
@@ -98,15 +97,21 @@ class Magnetometer
 
 	void Reset();
 
+	void SensorCorrectionsUpdate(bool force = false);
```
Comment (dagar) [inline src/lib/sensor_calibration/Magnetometer.hpp:100]: You'll need to call this periodically for all magnetometers somewhere in VehicleMagnetometer::Run().

Comment (mcsauder) [reply src/lib/sensor_calibration/Magnetometer.hpp:100]: Done. Thank you for noticing that.

File: src/drivers/magnetometer/isentek/ist8308/IST8308.hpp (hunk 1 above)
Comment (mcsauder) [reply src/drivers/magnetometer/isentek/ist8308/IST8308.hpp:105]: Yes. I'll have that pushed up shortly.

File: src/lib/sensor_calibration/Magnetometer.hpp (hunk 3 above)
Comment (mcsauder) [reply src/lib/sensor_calibration/Magnetometer.hpp:112]: Thanks.  That should be taken care of in SensorCorrectionsUpdate() just like the Accelerometer class and is now being called in VehicleMagnetometer per your review comment.  Thank you!

**Comment by mcsauder** [comment]
@dagar , comments addressed, thank you!  Squashed and rebased.

**Comment by mcsauder** [comment]
Flight tested [here](https://review.px4.io/plot_app?log=0366838d-1040-48b7-9907-91be32bb8c40), with ~20C drop in baro temp observed.

I bench tested to validate the coefficients are being applied/updated by entering bogus coefficients and observed the heading estimate drift wildly without any physical change in the vehicle orientation.

@dagar, any thoughts on what I could do about the flash size failures? Is there a mechanism to selectively exclude temp calibration from some builds?

File: boards/cuav/x7pro/default.px4board [hunk 5]
```diff
@@ -73,7 +73,6 @@ CONFIG_MODULES_RC_UPDATE=y
 CONFIG_MODULES_ROVER_POS_CONTROL=y
 CONFIG_MODULES_SENSORS=y
 CONFIG_MODULES_SIMULATION_SIMULATOR_SIH=y
-CONFIG_MODULES_TEMPERATURE_COMPENSATION=y
```
Comment (dagar) [inline boards/cuav/x7pro/default.px4board:76]: Can be reverted now?

File: boards/px4/fmu-v6c/default.px4board [hunk 6]
```diff
@@ -65,7 +65,6 @@ CONFIG_MODULES_RC_UPDATE=y
 CONFIG_MODULES_ROVER_POS_CONTROL=y
 CONFIG_MODULES_SENSORS=y
 CONFIG_MODULES_SIMULATION_SIMULATOR_SIH=y
-CONFIG_MODULES_TEMPERATURE_COMPENSATION=y
```
Comment (dagar) [inline boards/px4/fmu-v6c/default.px4board:68]: revert?

File: src/modules/commander/accelerometer_calibration.cpp [hunk 7]
```diff
@@ -177,7 +178,6 @@ static calibrate_return read_accelerometer_avg(float (&accel_avg)[MAX_ACCEL_SENS
 		{ORB_ID(sensor_accel), 0, 0},
 		{ORB_ID(sensor_accel), 0, 1},
 		{ORB_ID(sensor_accel), 0, 2},
-		{ORB_ID(sensor_accel), 0, 3},
```
Comment (dagar) [inline src/modules/commander/accelerometer_calibration.cpp:180]: I still think up to 4 accels is good (eg typical board with 3 internal + external connected).

**Comment by mcsauder** [comment]
I agree. But it adds fat. :/

File: src/modules/commander/accelerometer_calibration.cpp (hunk 7 above)
Comment (mcsauder) [reply src/modules/commander/accelerometer_calibration.cpp:180]: @dagar , would you like me to bring back the 4 accel/gyro instances and drop the mag/baro to 2 instance?

File: boards/cuav/x7pro/default.px4board (hunk 5 above)
Comment (mcsauder) [reply boards/cuav/x7pro/default.px4board]: Done. Thanks.

File: boards/px4/fmu-v6c/default.px4board (hunk 6 above)
Comment (mcsauder) [reply boards/px4/fmu-v6c/default.px4board]: Done. Thanks.

File: src/modules/commander/accelerometer_calibration.cpp (hunk 7 above)
Comment (mcsauder) [reply src/modules/commander/accelerometer_calibration.cpp:180]: @dagar, I went ahead and put the accel and gyro counts back to 4, mag and baro down to 2.  Commit pushed up.

**Comment by hamishwillee** [comment]
@mcsauder When this goes in, we will need to update http://docs.px4.io/main/en/advanced_config/sensor_thermal_calibration.html#thermal-calibration-and-compensation - I assume this will be in the release after PX4 v1.15, and that you will take care of initial update (might be as small as just removing the note up the top of that doc.)

**Pull request merged**